src/base.jl

```diff
@@ -1,21 +1,28 @@
+# help struct to keep internal stability
+struct AdaptTo{T} <: Function
+  to::T
+  AdaptTo(to) = new{Core.Typeof(to)}(to)
+end
+(f::AdaptTo)(x) = adapt(f.to, x)
+
 # predefined adaptors for working with types from the Julia standard library
 
-adapt_structure(to, xs::Union{Tuple,NamedTuple}) = map(x->adapt(to,x), xs)
+adapt_structure(to, xs::Union{Tuple,NamedTuple}) = map(AdaptTo(to), xs)
```

Comment:
Can you get the same benefit by using `Fix1`? Either directly in this `map`, or by defining `adapt(to) = Base.Fix1(adapt, to)`.

Reply:
Base.Fix1 is not stable for this use case except on master. We construct a Fix1{typeof(adapt),Datatype} before 1.9

Reply:
Not sure I follow, could you explain more?

Trying things locally, the `@inferred` tests seemed to pass with `Fix1`, although I was not very careful. But logically, this struct seems very similar to Fix1.

Reply:
Before 1.9, `Base.Fix1` is constructed with `Base.Fix1(f, x) = new{Core.Typeof(f),typeof(x)}(f, x)`
if `x` is a `Type`, our compiler would fail to infer `Base.Fix1(adapt, x)(y)` just like `(x->adapt(to,x))(y)`
```julia
julia> @code_warntype Base.Fix1(adapt, Array)(1:10)
MethodInstance for (::Base.Fix1{typeof(adapt), UnionAll})(::UnitRange{Int64})
  from (f::Base.Fix1)(y) in Base at operators.jl:1136
Arguments
  f::Base.Fix1{typeof(adapt), UnionAll}
  y::UnitRange{Int64}
Body::Union{UnitRange{Int64}, Vector}
1 ─ %1 = Base.getproperty(f, :f)::Core.Const(Adapt.adapt)
│   %2 = Base.getproperty(f, :x)::UnionAll
│   %3 = (%1)(%2, y)::Union{UnitRange{Int64}, Vector}
└──      return %3
```
Now on master, the second parameter has been further constrained to `Type{x}` in this case, so we have:
```julia
julia> @code_warntype Base.Fix1(adapt, Array)(1:10)
MethodInstance for (::Base.Fix1{typeof(adapt), Type{Array}})(::UnitRange{Int64})
  from (f::Base.Fix1)(y)
     @ Base operators.jl:1097
Arguments
  f::Base.Fix1{typeof(adapt), Type{Array}} # not a `UnionAll`
  y::UnitRange{Int64}
Body::Vector{Int64}
1 ─ %1 = Base.getproperty(f, :f)::Core.Const(Adapt.adapt)
│   %2 = Base.getproperty(f, :x)::Type{Array}
│   %3 = (%1)(%2, y)::Vector{Int64}
└──      return %3
```

Reply:
Ok, thanks, I didn't know Fix1 had this issue (and only tried on 1.9).

A possible work-around is just to make the right type by hand:
```
julia> macro new(T, args...)
            Expr(:new, T, args...) |> esc
       end;

julia> Adapt.adapt(::Type{T}) where T = @new(Base.Fix1{typeof(adapt),Type{T}}, adapt, T)

julia> adapt(Array)
(::Base.Fix1{typeof(adapt), Type{Array}}) (generic function with 1 method)

julia> @code_warntype adapt(Array)(1:10)
MethodInstance for (::Base.Fix1{typeof(adapt), Type{Array}})(::UnitRange{Int64})
  from (f::Base.Fix1)(y) in Base at operators.jl:1136
Arguments
  f::Base.Fix1{typeof(adapt), Type{Array}}
  y::UnitRange{Int64}
Body::Vector{Int64}
1 ─ %1 = Base.getproperty(f, :f)::Core.Const(Adapt.adapt)
│   %2 = Base.getproperty(f, :x)::Type{Array}
│   %3 = (%1)(%2, y)::Vector{Int64}
└──      return %3

julia> VERSION
v"1.7.0"
```

Reply:
Yes, `Expr(:new` is another solution for this.
If @maleadt is happy with the 1-arg `adapt`, I think we can switch to this and document it.

Reply:
A bit of a hack, I guess, but at least it's just applying a patch to get master's better behaviour. (I guess Adapt doesn't want to depend on Compat.jl, which would be another place that could live.)

The curried version sounds useful to me, for things like `model |> adapt(T)`. But could be discussed separately of course.

Reply:
Yeah the curried version seems nice and consistent with what's happening in Base.